src/KubernetesClient/Informers/Cache/VersionPartitionedSharedCache.cs

```diff
@@ -0,0 +1,294 @@
+        {
+            lock (_lock)
+            {
+                var view = new CacheView(this);
+                _views.Add(view);
+                return view;
+            }
+        }
+
+
+        private void Remove(TResource resource, CacheView originView)
+        {
+            var versionedKey = GetVersionKeyFor(resource);
+            Remove(versionedKey, originView);
+        }
+
+        private void Remove(VersionResourceKey versionedKey, CacheView originView)
+        {
+            var otherViewsTrackingResource = _views
+                .Except(new[] {originView})
+                .Any(x => x.TryGetValue(versionedKey.Key, out var resource) && _versionSelector(resource).Equals(versionedKey.Version));
+            if (!otherViewsTrackingResource)
+            {
+                Items.Remove(versionedKey);
+            }
+        }
+
+        private TResource GetOrAdd(TResource resource)
+        {
+            var key = GetVersionKeyFor(resource);
+            if (Items.TryGetValue(key, out var existingResource))
+                return existingResource;
+            Items.Add(key, resource);
+            return resource;
+        }
+
+        internal struct VersionResourceKey
+        {
+            public TKey Key;
+            public TVersion Version;
+        }
+
+        private VersionResourceKey GetVersionKeyFor(TResource resource)
+        {
+            return new VersionResourceKey() {Key = _keySelector(resource), Version = _versionSelector(resource)};
+        }
+
+        private class CacheView : ICache<TKey, TResource>
+        {
+            private readonly VersionPartitionedSharedCache<TKey, TResource, TVersion> _parent;
+            private readonly Dictionary<TKey, TResource> _items = new Dictionary<TKey, TResource>() ;
+
+            public long Version { get; set; }// = 1;
+
+            public CacheView(VersionPartitionedSharedCache<TKey, TResource, TVersion> parent)
+            {
+                _parent = parent;
+            }
+
+            public void Reset(IDictionary<TKey, TResource> newValues)
+            {
+                lock (_parent._lock)
+                {
+                    _items.Clear();
+                    foreach (var item in newValues)
+                    {
+                        _items.Add(item.Key, item.Value);
+                    }
+                }
+            }
+
+            public object SyncRoot => _parent._lock;
+            public ICacheSnapshot<TKey, TResource> Snapshot()
+            {
+                lock (_parent)
+                {
+                    return new SimpleCache<TKey, TResource>(this, Version);
+                }
+            }
+
+            public IEnumerator<KeyValuePair<TKey, TResource>> GetEnumerator()
+            {
+                lock (_parent._lock)
+                {
+                    return _items.ToList().GetEnumerator();
+                }
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
```

Comment:
Isn't this redundant with the method above?

Reply:
Part of interface of IDictionary, and no they are different.